### PR TITLE
Feat/error reporting

### PR DESCRIPTION
**Pull request by jonathanyeung**
# Pull Request Checklist

You can go to [dendron pull requests](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html) to see full details for items in this checklist.

### Quality Assurance
- [ ] add a [test](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#writing-tests) for the new feature
- [ ] make sure all the existing [tests](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#running-all-tests) pass
- [ ] do a spot check by running your feature with our [test workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace)
- [ ] after you submit your pull request, check the output of our [integration test](https://github.com/dendronhq/dendron/actions) and make sure all tests pass
  - NOTE: if you running mac/linux, check the windows output and vice versa if you are developing on windows

#### Special Cases
- [ ] if your tests changes an existing snaphot, make sure that snapshots are [updated](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#updating-test-snapshots)
- [ ] if you are adding a new language feature (graphically visible in vscode/preview/publishing), make sure that it is included in [test-workspace](https://wiki.dendron.so/notes/cb22bd36-d45a-4ffd-a31e-96c4b39cb7fb.html#test-workspace). We use this to manually inspect new changes and for auto regression testiing 

### Docs
- [x] Make sure that the PR title follows our [commit style](https://wiki.dendron.so/notes/adc39825-77a6-46cf-9c49-2642fcb4248e.html#commit-style)
- [x] Please summarize the feature or impact in 1-2 lines in the PR description
- [x] If your change reflects documentation changes, also submit a PR to [dendron-site](https://github.com/dendronhq/dendron-site) and mention the doc PR link in your current PR

Example PR Description
```markdown
# feat: capitalize all foos

This changes capitalizes all occurences of `foo` to `Foo` 

Docs PR: <URL_TO_DOCS_PR>
```

### First Time PR
- [ ] sign the [CLA](https://en.wikipedia.org/wiki/Contributor_License_Agreement) which will be prompted by our github bot after you submit the PR
- [ ] add your [discord](https://discord.gg/AE3NRw9) alias in the review so that we can give you the [horticulturalist](https://wiki.dendron.so/notes/7c00d606-7b75-4d28-b563-d75f33f8e0d7.html#horticulturalist) badge in our community

### Analytics
- [ ] if you are adding analytics related changes, make sure the [Telemetry](https://wiki.dendron.so/notes/84df871b-9442-42fd-b4c3-0024e35b5f3c.html) docs are updated
---
## QA 

Functionality should be covered by existing tests. 

Final manual testing with packaged build in progress.

### PR Description for Changelog: 
Error Analytics have been added to help us diagnose errors and improve the stability of Dendron. 

### Dendron-Site Changes

https://github.com/dendronhq/dendron-site/pull/194. This should also help explain some of the details of the PR.

### CODE REVIEW

File: packages/common-server/src/analytics.ts [hunk 1]
```diff
@@ -270,4 +272,50 @@ export class SegmentClient {
...
+  const dsn =
+    environment === "prod"
+      ? "https://bc206b31a30a4595a2efb31e8cc0c04e@o949501.ingest.sentry.io/5898219"
+      : undefined;
+
+  Sentry.init({
+    dsn,
+    defaultIntegrations: false,
+    tracesSampleRate: 1.0,
+    //TODO: respect user's telemetry settings on enabled.
```
Comment (jonathanyeung) [inline packages/common-server/src/analytics.ts:292]: TODO: Requires Fixing

File: packages/common-server/src/server.ts [hunk 2]
```diff
@@ -11,10 +12,29 @@ export class ExpressUtils {
...
+  static setResponse(
+    expressResponse: Response,
+    dendronResponse: RespV2<any>
+  ): void {
+    if (dendronResponse.error) {
+      expressResponse
+        .status(dendronResponse.error.code || StatusCodes.INTERNAL_SERVER_ERROR)
+        .json({ error: error2PlainObject(dendronResponse.error) });
+    } else {
+      //TODO: double check that this will convert properly.
```
Comment (jonathanyeung) [inline packages/common-server/src/server.ts:36]: TODO: Requires Check

File: .github/workflows/create-vsix-artifact.yml [hunk 3]
```diff
@@ -48,7 +49,10 @@ jobs:
         env: 
           GOOGLE_OAUTH_CLIENT_SECRET: ${{ secrets.GOOGLE_OAUTH_CLIENT_SECRET }}
           GOOGLE_OAUTH_CLIENT_ID: ${{ secrets.GOOGLE_OAUTH_CLIENT_ID }}
-        run: yarn build:patch:local:ci
+          SENTRY_AUTH_TOKEN: ${{ secrets.SENTRY_AUTH_TOKEN }}
+        run: |
+          echo "DENDRON_RELEASE_VERSION=`cat ./packages/plugin-core/package.json | jq ".version" -r | awk -F. -v OFS=. 'NF==1{print ++$NF}; NF>1{$NF=sprintf("%0*d", length($NF), ($NF+1)); print}'`" >> $GITHUB_ENV
```
Comment (kevinslin) [inline .github/workflows/create-vsix-artifact.yml:54]: heads up that as part of the build process we generate a `meta.json` file in `packages/plugin-core` that looks like the following. you can use this to get the release version

```json
{"@dendronhq/api-server":"0.58.3","@dendronhq/common-all":"0.58.3","@dendronhq/common-frontend":"0.58.3","@dendronhq/common-server":"0.58.3","@dendronhq/common-test-utils":"0.58.3","@dendronhq/dendron-11ty":"1.58.3","@dendronhq/dendron-cli":"0.58.3","dendron-design-system":"0.1.0","@dendronhq/dendron-next-server":"0.58.3","@dendronhq/engine-server":"0.58.3","@dendronhq/engine-test-utils":"0.58.3","generator-dendron":"0.1.0","@dendronhq/lsp-client-uber":"0.28.7-alpha.0","@dendronhq/lsp-server":"0.28.7-alpha.0","@dendronhq/plugin-core":"0.58.3","@dendronhq/pods-core":"0.58.3"}
```

**Review by kevinslin — CHANGES_REQUESTED** [review]
lgtm, minor comment about getting release version. 

1. can you update the [[Telemetry|dendron.ref.telemetry]] docs?
2. is there a reason this is still a draft?
3.  for future changes like this, can you break this apart into multiple iterations?
    1) make sure the thing works (eg. sentry error reporting, we can catch for one error)
    2) add exception handling everywhere
4. rebase from master to resolve conflicts

**Comment by jonathanyeung** [comment]
> lgtm, minor comment about getting release version.
> 
> 1. can you update the [[Telemetry|dendron.ref.telemetry]] docs?
> 2. is there a reason this is still a draft?
> 3. for future changes like this, can you break this apart into multiple iterations?
>    
>    1. make sure the thing works (eg. sentry error reporting, we can catch for one error)
>    2. add exception handling everywhere
> 4. rebase from master to resolve conflicts

Thanks for reviewing!

1. Updated the PR (https://github.com/dendronhq/dendron-site/pull/194) to include additional changes to the main telemetry page.
2. I had to do some additional testing on the .vsix first.  Moving out of draft mode now
3. Sounds good - iterate!
4. Done

**Review by kevinslin — APPROVED** [review]

**Pull request merged**